### PR TITLE
AOTAutograd: ensure traced tangent subclass metadata takes non-contiguous outputs into account

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/118596.

The issue was as follows:

(1) Whenever AOTAutograd sees an output that is non-contiguous, that it needs a tangent for, it forces the tangent that it generates to be contiguous during tracing

(2) However: if this tangent is a subclass, we need to generate code to flatten/unflatten the subclass at runtime.

(3) To do so, we use the metadata stashed here: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/schemas.py#L231

(4) However, this metadata was **wrong** - it was generated by inspecting the tangent, **before** we made the tangent contiguous.

The fix in this PR basically moves the logic make `traced_tangents` contiguous earlier, at the time that we first generate `ViewAndMutationMetadata`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118802
* #118670
* __->__ #118669
* #119947
* #118803

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

File: torch/_functorch/_aot_autograd/input_output_analysis.py [hunk 1]
```diff
@@ -48,7 +49,8 @@ def remove_dupe_metadata(
     other_traced_tangents = m.traced_tangents[num_data_mutations:]
     inp_traced_tangents = m.traced_tangents[:num_data_mutations]
     filtered_inp_traced_tangents = [
-        x
+        # See Note [Tangents must be contiguous]
+        coerce_tangent(x)
```
Comment (zou3519) [inline torch/_functorch/_aot_autograd/input_output_analysis.py:53]: why don't we need to coerce `other_traced_tangents`?

**Review by zou3519 — APPROVED** [review]
looks reasonable

File: torch/_functorch/_aot_autograd/input_output_analysis.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_functorch/_aot_autograd/input_output_analysis.py:53]: On second thought, I don't think i need this line at all. both of `inp_traced_tangents` and `other_traced_tangents` arte getting pulledfrom `m.traced_tangents`, which should have already gotten the `coerce_tangents` treatment at this point.

Ill remove and make sure that tests pass.

Comment (bdhirsh) [reply torch/_functorch/_aot_autograd/input_output_analysis.py:53]: update: it was indeed not necessary

**Pull request closed without merging**